### PR TITLE
Extract Method Refactoring - Multiple Exit Extraction#697

### PR DESCRIPTION
**Pull request by larjean89**
## What it does

- Added a new check to analyse the flow of data where multiple exits exist to prevent a method extraction which will lead to incorrect data flow analysis.
- Fixed issue https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/697)#697

## How to test

- Run test3000, test3001 and test3002 in ExtractMethodTests

## Author checklist

- [x] I have thoroughly tested my changes

- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by larjean89** [comment]
@jjohnstn is this pull request fine enough to be merged? If possible kindly check it out and let us know if there are any issues. Thank you so much for your time.

File: org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/binary/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/classes/
```
Comment (jjohnstn) [inline org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/binary/.gitignore:1]: This should not be necessary and the following 3 files should not be part of the PR.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/RefactoringCoreMessages.java [hunk 2]
```diff
@@ -514,6 +514,8 @@ public final class RefactoringCoreMessages extends NLS {
 
 	public static String ExtractInterfaceRefactoring_name;
 
+	public static String ExtractMethodAnalyzer_0;
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/RefactoringCoreMessages.java:517]: Change this to something meaningful like: ExtractMethodAnalyzer_multiple_exits

File: org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/ExtractMethodTestSetup.java [hunk 3]
```diff
@@ -169,5 +169,6 @@ public IPackageFragment getVarargsPackage() {
 	public IPackageFragment getFieldInitializerPackage() {
 		return fFieldInitializerPackage;
 	}
+
```
Comment (jjohnstn) [inline org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/ExtractMethodTestSetup.java:172]: Don't commit changes which are only formatting.

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Looks good overall.  Some minor comments to address.

File: org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/binary/.gitignore (hunk 1 above)
Comment (larjean89) [reply org.eclipse.jdt.ui.tests.refactoring/resources/InlineMethodWorkspace/TestCases/binary/.gitignore:1]: This is noted with thanks and will address this accordingly.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/RefactoringCoreMessages.java (hunk 2 above)
Comment (larjean89) [reply org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/RefactoringCoreMessages.java:517]: Thanks for the suggestion. Will address it.

File: org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/ExtractMethodTestSetup.java (hunk 3 above)
Comment (larjean89) [reply org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/ExtractMethodTestSetup.java:172]: That is noted with thanks and will address it.

**Comment by jjohnstn** [comment]
Hi @larjean89 Could you please rebase or remake this patch on current master and do a --force?  I tried but ran into conflict after conflict.  If you could remake your changes to ExtractMethodAnalyzer it would help.  I tried doing a rebase but there are so many commits that cause a conflict and it isn't obvious what each commit is doing.  If you can pare it down to one commit that would at least help if there is any other conflicts.

**Comment by larjean89** [comment]
@jjohnstn will do the rebase on the current master.
